### PR TITLE
Move wiki to developer guide in docs

### PR DESCRIPTION
**Pull request by dmentipl**
Addresses #1451

This PR adds content from the [IMPROVER wiki](https://github.com/metoppv/improver/wiki) to a "developer guide" section in the docs. Quoting from issue #1451:

> The content of the wiki is useful for developers. Moving it under source control means that changes to it require review, thus checking that the content is relevant and correct.

I have added all pages from the wiki to `docs/source` and changed the format from Markdown to reStructuredText, as expected by Sphinx, and fixed some minor formatting issues. I used [pandoc](https://pandoc.org/) to do this automatically after which I checked over the output manually. I have added a new "toctree" for the new content to `index.rst` under the heading "Developer Guide".

If this PR is accepted, I expect that we will remove the wiki.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: doc/source/Ticket-Creation-and-Definition-of-Ready.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+Have acceptance criteria.
+
+We also should make an effort to: \* Make it clear how we show the
+product owner it is done. \* Use the description box to provide more
+information as appropriate. \* Focus on requirements (necessary outputs,
+processes etc.) not solutions. Put suggested solutions in the comments
+on the tickets for discussion when the ticket is started. \* Use user
+story format “As … I want … so that …”. \* Be clear enough so there is a
+shared understanding of the ticket. \* List all dependencies and related
+work on the ticket.
```
Comment (bayliffe) [inline doc/source/Ticket-Creation-and-Definition-of-Ready.rst:103]: The bulleted lists in this file were a mix of hyphen and asterisk definitions; no reason for the inconsistency. The automatic translation has escaped all the asterisks, meaning these are no longer bulleted lists:
https://bens-improver-fork.readthedocs.io/en/wikitest/Ticket-Creation-and-Definition-of-Ready.html

File: doc/source/Ticket-Creation-and-Definition-of-Ready.rst [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+completed and what testing might be needed.
+
+A good reference for understanding what a user story is is here:
+http://www.romanpichler.com/blog/10-tips-writing-good-user-stories/
+
+Labelling issues
+----------------
+
+A number of labels are available for helping the categorise the
+available issues. Existing labels are: ### Type \* Administrative: The
```
Comment (bayliffe) [inline doc/source/Ticket-Creation-and-Definition-of-Ready.rst:33]: The Type heading is not encoded correctly.

File: doc/source/Running-at-your-site.rst [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+downloading the code from this repository, installing the required
+dependencies and then using either ``pip install`` (``pip install -e``)
+or ``python setup.py install`` (``python setup.py develop``). Note that
+``pip install`` will not work in an empty environment due to problems
+with installation of the dependency ``iris`` via pip.
+
+Required dependencies are listed in
+```environment_py36_iris22.yml`` <https://github.com/metoppv/improver/blob/master/envs/environment_py36_iris22.yml>`__
+and
+```environment_py37_iris24.yml`` <https://github.com/metoppv/improver/blob/master/envs/environment_py37_iris24.yml>`__,
```
Comment (bayliffe) [inline doc/source/Running-at-your-site.rst:28]: These display quite unpleasantly:
https://bens-improver-fork.readthedocs.io/en/wikitest/Running-at-your-site.html

File: doc/source/Guidance-for-Reviewing-Code.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+   3. **Ensure command line interface acceptance tests run and pass.
+      These must be ran on the desktop using bin/improver tests as they
+      are not run by travis**
+   4. **The Acceptance Criteria defined within the associated issue has
+      been met.**
+   5. **The criteria within the Definition of Done has been satisfied.**
+   6. **Ensure their testing is documented on the issue.**
+
+4. Main reviewers should post comments to the pull request to show that
+   they have completed: i, ii, iii, iv, v, vi.
```
Comment (bayliffe) [inline doc/source/Guidance-for-Reviewing-Code.rst:55]: The further indented bulleted list above has been relabelled with numbers rather than roman numerals, which means it no longer matches with this list of tasks to complete. Please can you use roman numerals (or letters) for the above else there is no way to distinguish between the top level numbered bullets and this inset list.

File: doc/source/Guidance-for-Reviewing-Code.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+   4. **Functionality**
+
+      -  Does the code do what is supposed to?
+      -  Are errors handled appropriately?
+
+   5. **Test coverage**
+
+      -  Do the tests provided cover the expected situations?
+      -  Are the tests understandable?
+      -  Have edge cases been considered?
```
Comment (bayliffe) [inline doc/source/Guidance-for-Reviewing-Code.rst:88]: Again, it would be nice if the inset list used different labels.

File: doc/source/Guidance-for-Reviewing-Code.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+7. If this is a second review, the developer should contact the Product
+   Owner prior to moving the issue into the “Done” column. The Scrum
+   Master can be consulted, if necessary.
+
+**Before merging:**
+-------------------
+
+Always rerun Travis before merging a PR. Do this by: 1. At the bottom of
+the PR, click ``show all checks`` and then ``details`` next to the
+Travis check. 2. Click ``Restart build``
```
Comment (bayliffe) [inline doc/source/Guidance-for-Reviewing-Code.rst:104]: 1 and 2 here should be a bulleted list to make it clearer; it's become inline in the translation process.

File: doc/source/Code-Style-Guide.rst [hunk 7]
```diff
@@ -0,0 +1,746 @@
...
+self-contained PRs for more manageable review. Reviewers should use
+their discretion in requesting PRs to be broken down if larger than
+this.
+
+Guidance (eg
+https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/)
+suggests that for reviews to be effective reviewers should: \* Take time
+over reviews: don’t review faster than 500 lines of code per hour \* Not
+review more than 400 lines of code at a time, or for more than one hour
+continuously
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:61]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 8]
```diff
@@ -0,0 +1,746 @@
...
+defined in its docstring to avoid partial information.
+
+Due to the use of `Sphinx <http://www.sphinx-doc.org/en/stable/>`__ for
+documentation building, a number of `docstring section
+headers <http://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html#docstring-sections>`__
+are supported. Favoured docstring section headers are: \* Args:
+Compulsory arguments. \* Keyword Args : Keyword arguments. \* Raises:
+Exceptions raised. \* Returns: Variables returned by the function or
+method. \* References: Link to available documentation. \* Warns:
+Warnings raised.
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:177]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+more examples.
+
+Variable types in CLIs
+~~~~~~~~~~~~~~~~~~~~~~
+
+Note that we use clize in the CLIs which uses type annotations at
+runtime, hence the need to define these clearly within the docstring.
+
+Within docstrings, when specifying a variable type, Python built-in data
+types can be used directly e.g. \* int \* float \* str \* bool
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:264]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 10]
```diff
@@ -0,0 +1,746 @@
...
+
+In order for variable types to link correctly within
+`readthedocs <http://improver.readthedocs.io/en/latest/?badge=latest>`__,
+the ``intersphinx_mapping`` needs to be updated to link to the
+documentation of the module where the variable type originates. For
+example: \* numpy.ndarray \* datatime.datetime The full name of the
+module is required, rather than an alias e.g. numpy rather than np.
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:271]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+they should live depends on how they will be used.
+
+**Case 1:** A function used in more than one module. - In this case the
+function should be located in a shared location e.g. utilities.py
+
+**Case 2:** A function used by several classes within one module. - The
+function should be kept outside of any one class, but within the module.
+
+**Case 3:** A function used only within a single class. - Should be kept
+within the class; as a static method if it makes no use of self.
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:495]: More bullets that has been inlined.
https://bens-improver-fork.readthedocs.io/en/wikitest/Code-Style-Guide.html#helper-functions-methods

File: doc/source/Code-Style-Guide.rst [hunk 12]
```diff
@@ -0,0 +1,746 @@
...
+To load the cubes, each cube argument will need a type. For a basic cube
+this will be ``cube: cli.inputcube``. If there is a default argument to
+go with the typed variable, spaces are required around the ``=`` for
+example ``weights: cli.inputcube = None``. There are other types which
+can be used such as: - the python standards - ``float`` - ``int`` -
+``bool`` - specific additions - ``cli.inputcube`` - Where a string is
+given which is a path to a cube to load - ``cli.inputjson`` - Where a
+string is given which is a path to a json file to load -
+``cli.comma_separated_list`` - This will convert the argument into that
+format and deal with error handling if no conversion is possible.
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:528]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 13]
```diff
@@ -0,0 +1,746 @@
...
+
+   from improver import cli
+   @cli.clizefy
+   def process(cube: cli.inputcube,
+               weights: cli.inputcube = None,
+               *,
+               coord_for_masking,
+               radius: float = None)
+
+The required arguments in this example are: - cube - coord_for_masking
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:556]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+`pytest <https://docs.pytest.org/en/latest/>`__.
+
+Unit Testing
+~~~~~~~~~~~~
+
+Add unit tests for functions and methods of classes where reasonable.
+
+You should add a unit test if: \* the logic in the function or method is
+not reasonably obvious from just looking at it, or \* the code is not
+otherwise covered by unit tests
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:577]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+
+Unit tests should: 1. Usually pass in representative inputs with the
+expected metadata and dimensions (consistent with what is enforced by
+the “load” module). 2. Use centralised test `cube set-up
+utilities <https://github.com/metoppv/improver/blob/master/improver/synthetic_data/set_up_test_cubes.py>`__
+where possible. 3. Consider the most likely uses of the plugin and
+ensure these are represented in the unit tests. 4. Consider possible
+edge cases e.g. cubes with different input dimensions. 5. Ensure the
+correct output is generated in good cases. 6. Ensure exceptions are
+raised as required for bad cases.
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:593]: Another bulleted list that has been inlined.

File: doc/source/Code-Style-Guide.rst [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+Each sub-directory must have a ``__init__.py``
+
+Tests for files containing functions external to classes should be in
+test files named for the source file:
+
+``e.g. utilities/spatial.py --> improver_tests/utilities/test_spatial.py``
+
+New unit test in IMPROVER should be written adhering to pytest style.
+These include: \* no test classes \* tests are written as functions \*
+test data is provided by fixtures
```
Comment (bayliffe) [inline doc/source/Code-Style-Guide.rst:616]: Another bulleted list that has been inlined.

**Review by bayliffe — CHANGES_REQUESTED** [review]
Thanks for this Daniel, it will look a lot nicer. Mostly bulleted lists that need sorting to make them render correctly.

File: doc/source/Ticket-Creation-and-Definition-of-Ready.rst (hunk 1 above)
Comment (dmentipl) [reply doc/source/Ticket-Creation-and-Definition-of-Ready.rst]: I've made unordered lists consistent (using `*`).

File: doc/source/Ticket-Creation-and-Definition-of-Ready.rst (hunk 2 above)
Comment (dmentipl) [reply doc/source/Ticket-Creation-and-Definition-of-Ready.rst]: Fixed.

File: doc/source/Running-at-your-site.rst (hunk 3 above)
Comment (dmentipl) [reply doc/source/Running-at-your-site.rst]: Unfortunately, it seems not possible (or least complicated) to get "monospace" font in a hyperlink. rST and Sphinx do have some quirks. I've made the links "regular" text with the hyperlink.

File: doc/source/Guidance-for-Reviewing-Code.rst (hunk 4 above)
Comment (dmentipl) [reply doc/source/Guidance-for-Reviewing-Code.rst:55]: I've fixed this. Plus it seems Sphinx requires very particular alignment.

File: doc/source/Guidance-for-Reviewing-Code.rst (hunk 5 above)
Comment (dmentipl) [reply doc/source/Guidance-for-Reviewing-Code.rst]: Fixed.

File: doc/source/Guidance-for-Reviewing-Code.rst (hunk 6 above)
Comment (dmentipl) [reply doc/source/Guidance-for-Reviewing-Code.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 7 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 8 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 9 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 10 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 11 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 12 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 13 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 14 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 15 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

File: doc/source/Code-Style-Guide.rst (hunk 16 above)
Comment (dmentipl) [reply doc/source/Code-Style-Guide.rst]: Fixed.

**Comment by dmentipl** [comment]
Thanks for the review @bayliffe. Sorry, I missed quite a bit of formatting issues caused by the pandoc conversion. Thanks for picking them up. I have fixed all the points you've raised, and corrected a few other issues I discovered. Hopefully it's ready to go now.

File: doc/source/Guidance-for-Reviewing-Code.rst (hunk 4 above)
Comment (bayliffe) [reply doc/source/Guidance-for-Reviewing-Code.rst]: Apparently there's a known issue with read-the-docs inheriting the format of the parent list for sublists in some cases:
https://github.com/readthedocs/sphinx_rtd_theme/issues/484

As such this change still renders on read-the-docs as a numbered list. Using letters seems to work both locally and on read-the-docs, so I've pushed up a change to do that.

https://bens-improver-fork.readthedocs.io/en/wikitest/Guidance-for-Reviewing-Code.html

**Review by bayliffe — APPROVED** [review]

File: doc/source/Guidance-for-Reviewing-Code.rst (hunk 4 above)
Comment (bayliffe) [reply doc/source/Guidance-for-Reviewing-Code.rst]: I'm happy with your changes, and my alterations on top are tiny. As such, I'm going to merge this and close down the wiki Thanks @dmentipl.

Comment (dmentipl) [reply doc/source/Guidance-for-Reviewing-Code.rst]: Thanks for looking into this and providing a solution 👍
